### PR TITLE
Detect when processes are not yielding.

### PR DESCRIPTION
**Pull request by floitsch**

### CODE REVIEW

File: src/scheduler.cc [hunk 1]
```diff
@@ -902,9 +910,20 @@ void Scheduler::tick(Locker& locker, int64 now) {
   for (SchedulerThread* thread : threads_) {
     Process* process = thread->interpreter()->process();
     if (process == null) continue;
+    uint64 us_since_preemption = now - process->run_timestamp();
+    if (process->signals() & Process::PREEMPT) {
+      // The process is already suppossed to preempt.
+      // Check whether it is stuck.
+      if (us_since_preemption > MAX_RUN_WITHOUT_PREEMPTION_US) {
+        FATAL("Process is not yielding\n");
```
Comment (kasperl) [inline src/scheduler.cc:918]: Maybe we can print the process better here? Even %p -- or perhaps better process->id() -- could be useful.

File: src/scheduler.cc [hunk 2]
```diff
@@ -885,7 +893,7 @@ void Scheduler::terminate_execution(Locker& locker, ExitState exit) {
 void Scheduler::tick(Locker& locker, int64 now) {
   tick_schedule(locker, now, true);
 
-  int first_non_empty_ready_queue = -1;
+  int first_non_empty_ready_queue = NUMBER_OF_READY_QUEUES;
```
Comment (kasperl) [inline src/scheduler.cc:896]: This makes the code from line 903-908 dead. Remove.

File: src/scheduler.cc (hunk 1 above)
Comment (kasperl) [inline src/scheduler.cc:918]:
```suggestion
        FATAL("Potential dead-lock detected in process %d\n", process->id());
```

**Review by kasperl — APPROVED** [review]
Looks good to me.

File: src/process.h [hunk 3]
```diff
@@ -259,6 +259,15 @@ class Process : public ProcessListFromProcessGroup::Element,
     return root_certificates_;
   }
 
+  /// Sets the timestamp for when this process was last preempted.
+  void set_run_timestamp(uint64 timestamp) {
+    run_timestamp_ = timestamp;
+  }
+
+  uint64 run_timestamp() {
```
Comment (kasperl) [inline src/process.h:267]: const

File: src/process.h [hunk 4]
```diff
@@ -318,6 +327,9 @@ class Process : public ProcessListFromProcessGroup::Element,
   HeapObject* null_;
 
   ResourceGroupListFromProcess resource_groups_;
+
+  uint64_t run_timestamp_ = 0;
```
Comment (kasperl) [inline src/process.h:331]:
```suggestion
  uint64 run_timestamp_ = 0;
```

File: src/scheduler.cc [hunk 5]
```diff
@@ -902,9 +910,20 @@ void Scheduler::tick(Locker& locker, int64 now) {
...
     if (process == null) continue;
+    uint64 us_since_preemption = now - process->run_timestamp();
+    if (process->signals() & Process::PREEMPT) {
+      // The process is already suppossed to preempt.
+      // Check whether it is stuck.
+      if (us_since_preemption > MAX_RUN_WITHOUT_PREEMPTION_US) {
+        FATAL("Process is not yielding\n");
+      } else {
+        continue;
+      }
```
Comment (kasperl) [inline src/scheduler.cc:921]:
```suggestion
      if (us_since_preemption <= MAX_RUN_WITHOUT_PREEMPTION_US) continue;
      FATAL("Process is not yielding\n");
```

File: src/process.h (hunk 3 above)
Comment (floitsch) [reply src/process.h]: done.

File: src/process.h (hunk 4 above)
Comment (floitsch) [reply src/process.h]: done.

File: src/scheduler.cc (hunk 2 above)
Comment (floitsch) [reply src/scheduler.cc:896]: yes.
I wanted to remove those lines anyway.
done.

File: src/scheduler.cc (hunk 1 above)
Comment (floitsch) [reply src/scheduler.cc]: done.

Comment (floitsch) [reply src/scheduler.cc]: done.

File: src/scheduler.cc (hunk 5 above)
Comment (floitsch) [reply src/scheduler.cc]: done.

**Pull request merged**